rfcs/0134-nix-store-layer.md

```diff
@@ -0,0 +1,257 @@
+
+### Guix
+
+Guix is further removed from Nix than Tvix + go-nix, and thus hints more at the end breadth of the design space yet to be explored.
+
+The store layer is the same, but the layers above, instead of being an alternative implementation of the Nix language, have a completely different design using Guile Scheme.
+The choice of language is just the tip of the iceberg here.
+More profoundly, they also have a more "library" than "interpreter" model where packages depend on Guix as a library, which talks to a small rump daemon.
+Guile sits far lower in their stack than the Nix language interpreter does; it is as if we rewrote some of our C++ into Nix language code, and Nix language code could do enough side effects to make that possible.
+
+The point of this discursion is to show that not only are radically different implementation of the same specification possible on either side of the store interface (what Tvix + go-nix aim for), but also radically different designs not at all trying to be compatible.
+
+All that said, below the store layer there is no difference in vision.
+Because our communities are so separate, it would be easy to come up with diverging versions of how derivations, store objects, etc. should work.
+That we have not done so I think is testament to the broad applicability of the Nix store design to many diverse groups of people with diverse goals.
+
+What I hope to do with Guix, then, is convene both projects to make their store store layers interoperate.
+Complementing the idea of a "marketplace of ideas" is when there is a certain design (like the Nix store layer), that is so broadly popular as to be a sort of "natural monopoly", that we should foster the most expansive and general idea of it as an exercise in coalition building and outreach.
+
+I do not expect Guix to be immediately sold on this plan, but as that larger project, I think it behooves us to take the first steps to build trust and coordination.
+Making a stand-alone Nix store executable demonstrates we are serious about layering and serious about standardizing that layer, and not just trying to get Guix users to use Nix instead.
+
+# Detailed design
+[design]: #detailed-design
+
+The goals motivated above a broken down into small steps that we can execute in isolation.
+This keeps the cost of this work initially lower, and generally reduces risk.
+
+## 1. The split itself
+
+Allow building a store-only version of Nix.
+This is a Nix executable that links `libstore` but not `libfetchers`, or `libexpr`.
+Plenty of commands like `nix daemon`, `nix log`, and the `nix store` sub-commands don't care about evaluation, fetching, or Flakes at all.
+[NixOS/nix#6182](https://github.com/NixOS/nix/issues/6182) implements this, splitting `libcmd` into two parts so the CLI code is reused.
+We will finish it off with as many commands as are reasonable to include, and merge it.
```

Comment:
Can we move the store-only parts of Nix into a separate drv output where non-dependency on the "full" output is enforced?

Reply:
Right now it would just be a separate derivation, but it would be nice to try to build both where possible.

Separate outputs might play funny with NixOS but I think we can. We can also just build libraries and exes in separate derivations.